chore: add yaml patch as override option

dannyrandall opened the pull request:
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

dannyrandall committed:
successfully creates the YAML patch directory

dannyrandall committed:
add readme.md

dannyrandall committed:
added cfn patch template

dannyrandall committed:
modified check for overrides

dannyrandall committed:
Merge branch 'mainline' into chore/override-patch

dannyrandall committed:
fix example patch

dannyrandall committed:
add more tests

dannyrandall committed:
add back err not exist for overrie lookup

dannyrandall committed:
rename for consistency

dannyrandall committed:
appease staticcheck

iamhopaul123 commented on internal/pkg/override/patch.go:
```diff
@@ -14,6 +15,18 @@ import (
 
 const jsonPointerSeparator = "/"
 
+func ScaffoldWithPatch(fs afero.Fs, dir string) error {
```
comment pls

dannyrandall replied:
there's gotta be some linter i can enable for this

iamhopaul123 commented on internal/pkg/override/override.go:
```diff
@@ -65,46 +65,28 @@ func (i Info) IsYAMLPatch() bool {
 // If path is a YAML patch document, then IsYAMLPatch evaluates to true.
```
is this comment still accurate?

iamhopaul123 commented on internal/pkg/override/override.go:
```diff
@@ -65,46 +65,28 @@ func (i Info) IsYAMLPatch() bool {
 
-type extension string
+	info, err := lookupYAMLPatch(path, fs)
+	if err == nil { // return yaml info if no error
+		return info, nil
+	}
 
-func (ext extension) isYAML() bool { return ext == ".yml" || ext == ".yaml" }
+	return lookupCDK(path, fs)
```
Hmm does that mean if any error occurs when executing `lookupYAMLPatch` we are going to swallow that error? Like users are using yaml patch but the file is misconfigured but we show them error like `"cdk.json" does not exist under...` 🤔

dannyrandall replied:
mmm fair point. how about this - we'll only test for the existence of a file here. any misconfiguration error will happen later, when attempting to apply the patches (just like CDK). how does that sound?

iamhopaul123 replied:
:+1:

iamhopaul123 commented on internal/pkg/template/overrides.go:
```diff
@@ -130,3 +132,7 @@ func (t *Template) WalkOverridesCDKDir(resources []CFNResource, fn WalkDirFunc)
 	))
 }
+
+func (t *Template) WalkOverridesPatchDir(fn WalkDirFunc) error {
```
comment pls

efekarakus commented on internal/pkg/cli/override.go:
```diff
@@ -20,13 +20,15 @@ import (
 const (
 	// IaC options for overrides.
 	cdkIaCTool = "cdk"
+	yamlPatch  = "CloudFormation YAML patches"
```
Can we do:
```suggestion
	yamlPatch  = "yamlpatch"
```

And update the help menu here:
https://github.com/aws/copilot-cli/blob/318be13f9280352de52983470abd23cd3ccf9039/internal/pkg/cli/override.go#L148-L151

with an explanation of `yamlpatch`, for example:
```go
	help := `The AWS Cloud Development Kit (CDK) lets you override templates using
the expressive power of programming languages.
This option is recommended for users that need to override several resources.
To learn more about the CDK: https://aws.github.io/copilot-cli/docs/developing/overrides/cdk/

CloudFormation YAML patches is recommended for users that need to override
a handful resources or do not want to depend on any other tool.
To learn more about CFN yaml patches: https://aws.github.io/copilot-cli/docs/developing/overrides/yamlpatch/
`
```

dannyrandall replied:
![](https://media.tenor.com/s1w62ls6xlQAAAAC/princess-bride.gif)

efekarakus commented on internal/pkg/override/patch.go:
```diff
@@ -78,6 +92,7 @@ func (p *Patch) Override(body []byte) ([]byte, error) {
 }
 
 func unmarshalPatches(path string, fs afero.Fs) ([]yamlPatch, error) {
+	path = filepath.Join(path, "cfn.patches.yml")
```
nit: can we move this to a const in this file? It seems important enough to highlight it

dannyrandall replied:
![](https://media.tenor.com/s1w62ls6xlQAAAAC/princess-bride.gif)

iamhopaul123 replied:
lmao

efekarakus commented on internal/pkg/template/templates/overrides/yamlpatch/README.md:
```diff
@@ -0,0 +1,17 @@
+operations are supported by Copilot.
+
+Patches are applied in the order specified in the file.
+
```
```suggestion
## Further material

To view examples and an explanation of how YAML patches work, check out the [documentation](https://aws.github.io/copilot-cli/docs/developing/overrides/yamlpatch).

```

dannyrandall replied:
see what you think - i'll highlight our page over the RFC?

efekarakus commented on internal/pkg/template/templates/overrides/yamlpatch/cfn.patches.yml:
```diff
@@ -0,0 +1,15 @@
+# Delete the task role resource
+- op: remove
+  path: /Resources/TaskRole
+
+# Add a service connect alias
+- op: add
+  path: /Resources/Service/Properties/ServiceConnectConfiguration/Services/0/ClientAliases/-
+  value:
+    Port: !Ref TargetPort
+    DnsName: yamlpatchiscool
+
+# Replace the task role in the task definition
+- op: replace
+  path: /Resources/TaskDefinition/Properties/TaskRoleArn
+  value: arn:aws:iam::123456789012:role/MyTaskRole
```
Should we comment out the entire file, so that when a customer runs `copilot deploy` immediately after the `override` command we don't suddenly delete the task role 😝

dannyrandall replied:
it's a feature not a bug

efekarakus reviewed (commented):
Weeee 🚀 looks awesome, consider marking the PR as a `feat`

dannyrandall committed:
fb from @iamhopaul123: fix comments

dannyrandall committed:
fb from @efekarakus:
- `const` for patch file name
- point to copilot docs in patch readme
- comment out example patches

dannyrandall committed:
move lookup error of parsing yaml patches to `Override()`

dannyrandall committed:
fix test failure

dannyrandall committed:
missed a test

dannyrandall committed:
Merge branch 'mainline' into chore/override-patch

efekarakus committed:
Merge branch 'mainline' into chore/override-patch

mergify[bot] committed:
Merge branch 'mainline' into chore/override-patch
